### PR TITLE
benchmark and optimize httpevents

### PR DESCRIPTION
**Pull request by achille-roussel**
I noticed logging an HTTP request was quite expensive in both time and memory, so I spent a bit of time to add a few optimizations.

Because some of the optimizations rely on unsafe assumptions about the memory layout of interfaces I made those opt-in, the program may enable them by setting `events.EnableUnsafeOptimizations = true`.

Here are the comparison of running the benchmarks before and after:
```
benchmark                 old ns/op     new ns/op     delta
BenchmarkRequestLog-4     1315          796           -39.47%
BenchmarkHandler-4        1020          918           -10.00%

benchmark                 old allocs     new allocs     delta
BenchmarkRequestLog-4     11             0              -100.00%
BenchmarkHandler-4        1              0              -100.00%

benchmark                 old bytes     new bytes     delta
BenchmarkRequestLog-4     296           0             -100.00%
BenchmarkHandler-4        48            0             -100.00%
```

Please take a look and let me know if something should be changed.

### CODE REVIEW

**Comment by f2prateek** [comment]
From an API standpoint, I don't 100% understand the negative implications of enabling this flag, and I think most people will not as well. So I'm worried people will enable this without understanding the consequences.

**Comment by achille-roussel** [comment]
OK I'll improve the documentation.

**Comment by achille-roussel** [comment]
@f2prateek I thought more about your comment and I agree that giving the developer the ability to control when to turn on/off the optimizations is asking for trouble, unless you have a deep understanding of the memory layout used by the compiler you're likely to do the wrong thing. On top of that, it's something you have to do in every program and therefore has a very high maintenance cost to get right.

That being said, for a given version of Go we can for sure know what kind of code the compiler is generating. I changed the approach to use conditional compilation instead, so the developer doesn't need to get involved, and if the compiler ever make changes to how interfaces are represented in memory we can always either disable the optimization for that version of Go, or rewrite them to be compatible with the new version and use the same conditional compilation approach to select the code that gets generated.

I kind of like this approach and it gave me an idea for another optimization that prevents the dynamic memory allocations that are generated for each argument passed to the variable argument list of `Debug(string, ...interface{})` and `Log(string, ...interface{})`.
We now have code which is fully free of dynamic memory allocations caused by generated calls to `runtime.convT2E` (@yields enjoy!)
```
benchmark                    old ns/op     new ns/op     delta
BenchmarkLogger/inline-4     432           350           -18.98%

benchmark                    old allocs     new allocs     delta
BenchmarkLogger/inline-4     3              0              -100.00%

benchmark                    old bytes     new bytes     delta
BenchmarkLogger/inline-4     24            0             -100.00%
```

Please take a look again and let me know if something should be changed.

**Comment by achille-roussel** [comment]
@f2prateek 
@yields 
@abraithwaite 

Is there anything more to add or remove in this PR?

File: event_unsafe.go [hunk 1]
```diff
@@ -0,0 +1,39 @@
+// +build go1.5
```
Comment (f2prateek) [inline event_unsafe.go]: Is 1.5 intentional here and in event_safe.go? We're already on 1.8 I think

File: httpevents/request_safe.go [hunk 2]
```diff
@@ -0,0 +1,15 @@
+// +build !go1.5
+
+package httpevents
+
+func bytesToStringNonEmpty(b []byte) string {
```
Comment (f2prateek) [inline httpevents/request_safe.go]: Does this need to be it's own function? bytesToString is functionally the same I think (with an extra if check).

Comment (achille-roussel) [inline httpevents/request_safe.go]: Yes, but they live in different packages.

File: event_unsafe.go (hunk 1 above)
Comment (achille-roussel) [inline event_unsafe.go]: this means 1.5+ => https://golang.org/pkg/go/build/#hdr-Build_Constraints

**Comment by achille-roussel** [comment]
Thanks for the review!

**Pull request merged**